File: python/microsoft-agents-protocol-llm/microsoft_agents_protocol_llm/openai_protocol_client.py
"""
OpenAI implementation of Protocol-Native LLM client.
"""

import uuid
from typing import AsyncIterator, Optional, Any
from openai import AsyncOpenAI
from microsoft_agents_protocol.models import (
    ChatMessage,
    AgentMessage,
    SystemMessage,
    UserMessage,
    ToolMessage,
    TextContent,
    ImageContent,
    FunctionCallContent,
    FunctionResultContent,
    AIContent,
)

from .protocol_llm_client import (
    ProtocolLLMClient,
    AgentMessageDelta,
    DeltaType,
    LLMProviderInfo,
    ToolDefinition,
)


class OpenAIProtocolClient(ProtocolLLMClient):
    """
    OpenAI implementation of IProtocolLLMClient that returns Agent Protocol types directly.
    """

    def __init__(
        self,
        api_key: str,
        model: str = "gpt-4o",
        base_url: Optional[str] = None,
        temperature: float = 1.0,
        max_tokens: Optional[int] = None,
        seed: Optional[int] = None,
    ):
        """
        Creates a new OpenAI protocol client.

        Args:
            api_key: OpenAI API key
            model: Model identifier (default: gpt-4o)
            base_url: Optional custom base URL (for Azure, Foundry, etc.)
            temperature: Temperature for generation (0.0 = deterministic, 2.0 = very random)
            max_tokens: Maximum number of tokens to generate
            seed: Random seed for deterministic generation
        """
        self._model = model
        self._temperature = temperature
        self._max_tokens = max_tokens
        self._seed = seed

        client_kwargs = {"api_key": api_key}
        if base_url:
            client_kwargs["base_url"] = base_url

        self._client = AsyncOpenAI(**client_kwargs)

    @property
    def provider_info(self) -> LLMProviderInfo:
        return LLMProviderInfo(
            provider="OpenAI",
            model=self._model,
            supports_streaming=True,
            supports_function_calling=True,
            supports_vision="vision" in self._model or "4o" in self._model or "gpt-4" in self._model,
            supports_multimodal="4o" in self._model,
        )

    async def generate(
        self,
        conversation_history: list[ChatMessage],
        available_tools: Optional[list[ToolDefinition]] = None,
    ) -> AgentMessage:
        """Generate a response using Agent Protocol message types."""
        openai_messages = self._convert_to_openai_messages(conversation_history)
        openai_tools = self._convert_to_openai_tools(available_tools) if available_tools else None

        completion_kwargs = {
            "model": self._model,
            "messages": openai_messages,
            "temperature": self._temperature,
        }

        if openai_tools:
            completion_kwargs["tools"] = openai_tools

        if self._max_tokens:
            completion_kwargs["max_tokens"] = self._max_tokens

        if self._seed is not None:
            completion_kwargs["seed"] = self._seed

        completion = await self._client.chat.completions.create(**completion_kwargs)

        return self._convert_to_agent_message(completion)

    async def stream(
        self,
        conversation_history: list[ChatMessage],
        available_tools: Optional[list[ToolDefinition]] = None,
    ) -> AsyncIterator[AgentMessageDelta]:
        """Stream a response using Agent Protocol message types."""
        openai_messages = self._convert_to_openai_messages(conversation_history)
        openai_tools = self._convert_to_openai_tools(available_tools) if available_tools else None

        completion_kwargs = {
            "model": self._model,
            "messages": openai_messages,
            "temperature": self._temperature,
            "stream": True,
        }

        if openai_tools:
            completion_kwargs["tools"] = openai_tools

        if self._max_tokens:
            completion_kwargs["max_tokens"] = self._max_tokens

        if self._seed is not None:
            completion_kwargs["seed"] = self._seed

        stream = await self._client.chat.completions.create(**completion_kwargs)

        message_id = f"msg_{uuid.uuid4().hex}"
        text_buffer = ""
        tool_call_buffers = {}

        async for chunk in stream:
            if not chunk.choices:
                continue

            delta = chunk.choices[0].delta

            # Start of message
            if delta.role:
                yield AgentMessageDelta(
                    message_id=message_id, type=DeltaType.MESSAGE_START
                )

            # Text content
            if delta.content:
                text_buffer += delta.content
                yield AgentMessageDelta(
                    message_id=message_id,
                    type=DeltaType.TEXT_DELTA,
                    content=TextContent(text=text_buffer),
                )

            # Tool calls
            if delta.tool_calls:
                for tool_call in delta.tool_calls:
                    index = tool_call.index

                    if index not in tool_call_buffers:
                        tool_call_buffers[index] = {
                            "call_id": tool_call.id or f"call_{uuid.uuid4().hex}",
                            "name": tool_call.function.name if tool_call.function else "",
                            "arguments": "",
                        }

                        yield AgentMessageDelta(
                            message_id=message_id,
                            type=DeltaType.TOOL_CALL_START,
                            tool_call={
                                "call_id": tool_call_buffers[index]["call_id"],
                                "name": tool_call_buffers[index]["name"],
                            },
                        )

                    if tool_call.function and tool_call.function.arguments:
                        tool_call_buffers[index]["arguments"] += tool_call.function.arguments

                        yield AgentMessageDelta(
                            message_id=message_id,
                            type=DeltaType.TOOL_CALL_DELTA,
                            tool_call={
                                "call_id": tool_call_buffers[index]["call_id"],
                                "name": tool_call_buffers[index]["name"],
                                "arguments": tool_call_buffers[index]["arguments"],
                            },
                        )

            # End of message
            if chunk.choices[0].finish_reason:
                # Complete any pending tool calls
                for builder in tool_call_buffers.values():
                    yield AgentMessageDelta(
                        message_id=message_id,
                        type=DeltaType.TOOL_CALL_COMPLETE,
                        tool_call=builder,
                    )

                yield AgentMessageDelta(
                    message_id=message_id,
                    type=DeltaType.MESSAGE_COMPLETE,
                    is_complete=True,
                )

    def _convert_to_agent_message(self, completion: Any) -> AgentMessage:
        """Convert OpenAI completion to Agent Protocol message."""
        contents = []

        message = completion.choices[0].message

        # Text content
        if message.content:
            contents.append(TextContent(text=message.content))

        # Tool calls
        if message.tool_calls:
            for tool_call in message.tool_calls:
                contents.append(
                    FunctionCallContent(
                        call_id=tool_call.id,
                        name=tool_call.function.name,
                        arguments=tool_call.function.arguments,
                    )
                )

        return AgentMessage(
            message_id=f"msg_{completion.id or uuid.uuid4().hex}", contents=contents
        )

    def _convert_to_openai_messages(
        self, protocol_messages: list[ChatMessage]
    ) -> list[dict[str, Any]]:
        """Convert Agent Protocol messages to OpenAI format."""
        openai_messages = []

        for msg in protocol_messages:
            if isinstance(msg, SystemMessage):
                text_content = next((c.text for c in msg.contents if isinstance(c, TextContent)), None)
                if text_content:
                    openai_messages.append({"role": "system", "content": text_content})

            elif isinstance(msg, UserMessage):
                # Handle multimodal content
                content_parts = []
                for content in msg.contents:
                    if isinstance(content, TextContent):
                        content_parts.append({"type": "text", "text": content.text})
                    elif isinstance(content, ImageContent) and content.image_url:
                        content_parts.append(
                            {"type": "image_url", "image_url": {"url": content.image_url}}
                        )

                if len(content_parts) == 1 and content_parts[0]["type"] == "text":
                    openai_messages.append(
                        {"role": "user", "content": content_parts[0]["text"]}
                    )
                elif content_parts:
                    openai_messages.append({"role": "user", "content": content_parts})

            elif isinstance(msg, AgentMessage):
                openai_msg = {"role": "assistant"}

                text_parts = [c.text for c in msg.contents if isinstance(c, TextContent)]
                if text_parts:
                    openai_msg["content"] = " ".join(text_parts)

                tool_calls = [c for c in msg.contents if isinstance(c, FunctionCallContent)]
                if tool_calls:
                    openai_msg["tool_calls"] = [
                        {
                            "id": tc.call_id or f"call_{uuid.uuid4().hex}",
                            "type": "function",
                            "function": {"name": tc.name or "", "arguments": tc.arguments or "{}"},
                        }
                        for tc in tool_calls
                    ]

                openai_messages.append(openai_msg)

            elif isinstance(msg, ToolMessage):
                for content in msg.contents:
                    if isinstance(content, FunctionResultContent):
                        openai_messages.append(
                            {
                                "role": "tool",
                                "tool_call_id": content.call_id or f"call_{uuid.uuid4().hex}",
                                "content": content.result or "",
                            }
                        )

        return openai_messages

    def _convert_to_openai_tools(
        self, protocol_tools: list[ToolDefinition]
    ) -> list[dict[str, Any]]:
        """Convert Agent Protocol tools to OpenAI format."""
        return [
            {
                "type": "function",
                "function": {
                    "name": tool.function.name,
                    "description": tool.function.description or "",
                    "parameters": tool.function.parameters or {},
                },
            }
            for tool in protocol_tools
        ]

File: dotnet/src/Microsoft.Agents.Protocol.Model.OpenAI/OpenAIProtocolClient.cs
using Microsoft.Agents;
using Microsoft.Agents.Protocol.Model;
using OpenAI;
using OpenAI.Chat;
using System.Runtime.CompilerServices;
using System.Text.Json;
using OpenAIChatMessage = OpenAI.Chat.ChatMessage;
using ProtocolChatMessage = Microsoft.Agents.ChatMessage;

namespace Microsoft.Agents.Protocol.Model.OpenAI;

/// <summary>
/// OpenAI implementation of IProtocolLLMClient that returns Agent Protocol types directly.
/// </summary>
public class OpenAIProtocolClient : IProtocolLLMClient
{
    private readonly ChatClient _client;
    private readonly string _model;
    private readonly OpenAIProtocolClientOptions _options;

    /// <summary>
    /// Creates a new OpenAI protocol client.
    /// </summary>
    /// <param name="apiKey">OpenAI API key</param>
    /// <param name="model">Model identifier (default: gpt-4o)</param>
    /// <param name="options">Optional client configuration</param>
    public OpenAIProtocolClient(
        string apiKey,
        string model = "gpt-4o",
        OpenAIProtocolClientOptions? options = null)
    {
        ArgumentNullException.ThrowIfNull(apiKey);
        ArgumentNullException.ThrowIfNull(model);

        _model = model;
        _options = options ?? new OpenAIProtocolClientOptions();

        var openAIClient = new OpenAIClient(apiKey);
        _client = openAIClient.GetChatClient(_model);
    }

    /// <summary>
    /// Creates a new OpenAI protocol client with custom base URL (for Azure, Foundry, etc.).
    /// </summary>
    /// <param name="apiKey">API key</param>
    /// <param name="baseUrl">Custom base URL</param>
    /// <param name="model">Model identifier</param>
    /// <param name="options">Optional client configuration</param>
    public OpenAIProtocolClient(
        string apiKey,
        Uri baseUrl,
        string model,
        OpenAIProtocolClientOptions? options = null)
    {
        ArgumentNullException.ThrowIfNull(apiKey);
        ArgumentNullException.ThrowIfNull(baseUrl);
        ArgumentNullException.ThrowIfNull(model);

        _model = model;
        _options = options ?? new OpenAIProtocolClientOptions();

        var openAIClient = new OpenAIClient(new System.ClientModel.ApiKeyCredential(apiKey), new OpenAIClientOptions
        {
            Endpoint = baseUrl
        });
        _client = openAIClient.GetChatClient(_model);
    }

    public LLMProviderInfo ProviderInfo => new()
    {
        Provider = "OpenAI",
        Model = _model,
        SupportsStreaming = true,
        SupportsFunctionCalling = true,
        SupportsVision = _model.Contains("vision") || _model.Contains("4o") || _model.Contains("gpt-4"),
        SupportsMultimodal = _model.Contains("4o")
    };

    public async Task<AgentMessage> GenerateAsync(
        List<ProtocolChatMessage> conversationHistory,
        ToolDefinition[]? availableTools = null,
        CancellationToken cancellationToken = default)
    {
        var openAIMessages = ConvertToOpenAIMessages(conversationHistory);
        var openAITools = ConvertToOpenAITools(availableTools);

        var chatOptions = new ChatCompletionOptions();

        if (openAITools != null && openAITools.Count > 0)
        {
            foreach (var tool in openAITools)
            {
                chatOptions.Tools.Add(tool);
            }
        }

        chatOptions.Temperature = (float)_options.Temperature;

        if (_options.MaxTokens.HasValue)
        {
            chatOptions.MaxOutputTokenCount = _options.MaxTokens.Value;
        }

        if (_options.TopP.HasValue)
        {
            chatOptions.TopP = (float)_options.TopP.Value;
        }

        if (_options.FrequencyPenalty.HasValue)
        {
            chatOptions.FrequencyPenalty = (float)_options.FrequencyPenalty.Value;
        }

        if (_options.PresencePenalty.HasValue)
        {
            chatOptions.PresencePenalty = (float)_options.PresencePenalty.Value;
        }

        var completion = await _client.CompleteChatAsync(openAIMessages, chatOptions, cancellationToken);

        return ConvertToAgentMessage(completion.Value);
    }

    public async IAsyncEnumerable<AgentMessageDelta> StreamAsync(
        List<ProtocolChatMessage> conversationHistory,
        ToolDefinition[]? availableTools = null,
        [EnumeratorCancellation] CancellationToken cancellationToken = default)
    {
        var openAIMessages = ConvertToOpenAIMessages(conversationHistory);
        var openAITools = ConvertToOpenAITools(availableTools);

        var chatOptions = new ChatCompletionOptions();

        if (openAITools != null && openAITools.Count > 0)
        {
            foreach (var tool in openAITools)
            {
                chatOptions.Tools.Add(tool);
            }
        }

        chatOptions.Temperature = (float)_options.Temperature;

        if (_options.MaxTokens.HasValue)
        {
            chatOptions.MaxOutputTokenCount = _options.MaxTokens.Value;
        }

        if (_options.TopP.HasValue)
        {
            chatOptions.TopP = (float)_options.TopP.Value;
        }

        var streamingUpdates = _client.CompleteChatStreamingAsync(openAIMessages, chatOptions, cancellationToken);

        var messageId = $"msg_{Guid.NewGuid():N}";
        var textBuffer = "";
        var toolCallBuffers = new Dictionary<int, ToolCallBuilder>();

        await foreach (var update in streamingUpdates.ConfigureAwait(false))
        {
            cancellationToken.ThrowIfCancellationRequested();

            // Start of message
            if (update.Role.HasValue)
            {
                yield return new AgentMessageDelta
                {
                    MessageId = messageId,
                    Type = DeltaType.MessageStart
                };
            }

            // Text content
            foreach (var contentPart in update.ContentUpdate)
            {
                if (contentPart.Text != null)
                {
                    textBuffer += contentPart.Text;

                    yield return new AgentMessageDelta
                    {
                        MessageId = messageId,
                        Type = DeltaType.TextDelta,
                        Content = new TextContent { Text = textBuffer }
                    };
                }
            }

            // Tool calls
            foreach (var toolCallUpdate in update.ToolCallUpdates)
            {
                var index = toolCallUpdate.Index;

                if (!toolCallBuffers.ContainsKey(index))
                {
                    toolCallBuffers[index] = new ToolCallBuilder
                    {
                        CallId = toolCallUpdate.ToolCallId ?? $"call_{Guid.NewGuid():N}",
                        FunctionName = toolCallUpdate.FunctionName ?? ""
                    };

                    yield return new AgentMessageDelta
                    {
                        MessageId = messageId,
                        Type = DeltaType.ToolCallStart,
                        ToolCall = new FunctionCallContent
                        {
                            CallId = toolCallBuffers[index].CallId,
                            Name = toolCallBuffers[index].FunctionName
                        }
                    };
                }

                if (toolCallUpdate.FunctionArgumentsUpdate.ToMemory().Length > 0)
                {
                    toolCallBuffers[index].ArgumentsBuffer += toolCallUpdate.FunctionArgumentsUpdate.ToString();

                    yield return new AgentMessageDelta
                    {
                        MessageId = messageId,
                        Type = DeltaType.ToolCallDelta,
                        ToolCall = new FunctionCallContent
                        {
                            CallId = toolCallBuffers[index].CallId,
                            Name = toolCallBuffers[index].FunctionName,
                            Arguments = toolCallBuffers[index].ArgumentsBuffer
                        }
                    };
                }
            }

            // End of message
            if (update.FinishReason.HasValue)
            {
                // Complete any pending tool calls
                foreach (var builder in toolCallBuffers.Values)
                {
                    yield return new AgentMessageDelta
                    {
                        MessageId = messageId,
                        Type = DeltaType.ToolCallComplete,
                        ToolCall = new FunctionCallContent
                        {
                            CallId = builder.CallId,
                            Name = builder.FunctionName,
                            Arguments = builder.ArgumentsBuffer
                        }
                    };
                }

                yield return new AgentMessageDelta
                {
                    MessageId = messageId,
                    Type = DeltaType.MessageComplete,
                    IsComplete = true
                };
            }
        }
    }

    private AgentMessage ConvertToAgentMessage(ChatCompletion openAICompletion)
    {
        var contents = new List<AIContent>();
        var assistantMessage = openAICompletion.Content;

        // Text content
        foreach (var contentPart in assistantMessage)
        {
            if (contentPart.Text != null)
            {
                contents.Add(new TextContent { Text = contentPart.Text });
            }
        }

        // Function/tool calls
        foreach (var toolCall in openAICompletion.ToolCalls)
        {
            contents.Add(new FunctionCallContent
            {
                CallId = toolCall.Id,
                Name = toolCall.FunctionName,
                Arguments = toolCall.FunctionArguments.ToString()
            });
        }

        return new AgentMessage
        {
            MessageId = $"msg_{openAICompletion.Id ?? Guid.NewGuid().ToString("N")}",
            Contents = contents
        };
    }

    private List<OpenAIChatMessage> ConvertToOpenAIMessages(List<ProtocolChatMessage> protocolMessages)
    {
        var openAIMessages = new List<OpenAIChatMessage>();

        foreach (var msg in protocolMessages)
        {
            switch (msg)
            {
                case SystemMessage system:
                    var systemText = system.Contents.OfType<TextContent>().FirstOrDefault()?.Text;
                    if (systemText != null)
                    {
                        openAIMessages.Add(new SystemChatMessage(systemText));
                    }
                    break;

                case UserMessage user:
                    var userContents = new List<ChatMessageContentPart>();
                    foreach (var content in user.Contents)
                    {
                        if (content is TextContent textContent)
                        {
                            userContents.Add(ChatMessageContentPart.CreateTextPart(textContent.Text));
                        }
                        else if (content is ImageContent imageContent)
                        {
                            // Support image URLs
                            if (!string.IsNullOrEmpty(imageContent.Uri))
                            {
                                userContents.Add(ChatMessageContentPart.CreateImagePart(new Uri(imageContent.Uri)));
                            }
                        }
                    }

                    if (userContents.Count == 1 && userContents[0].Kind == ChatMessageContentPartKind.Text)
                    {
                        openAIMessages.Add(new UserChatMessage(userContents[0].Text));
                    }
                    else if (userContents.Count > 0)
                    {
                        openAIMessages.Add(new UserChatMessage(userContents));
                    }
                    break;

                case AgentMessage agent:
                    var textParts = agent.Contents.OfType<TextContent>().ToList();
                    var toolCalls = agent.Contents.OfType<FunctionCallContent>().ToList();

                    if (toolCalls.Any())
                    {
                        var assistantMessage = new AssistantChatMessage(
                            toolCalls.Select(tc => ChatToolCall.CreateFunctionToolCall(
                                tc.CallId ?? $"call_{Guid.NewGuid():N}",
                                tc.Name ?? "",
                                BinaryData.FromString(tc.Arguments ?? "{}")
                            )).ToList()
                        );

                        if (textParts.Any())
                        {
                            assistantMessage.Content.Add(ChatMessageContentPart.CreateTextPart(
                                string.Join(" ", textParts.Select(t => t.Text))
                            ));
                        }

                        openAIMessages.Add(assistantMessage);
                    }
                    else if (textParts.Any())
                    {
                        openAIMessages.Add(new AssistantChatMessage(
                            string.Join(" ", textParts.Select(t => t.Text))
                        ));
                    }
                    break;

                case ToolMessage tool:
                    foreach (var content in tool.Contents.OfType<FunctionResultContent>())
                    {
                        openAIMessages.Add(new ToolChatMessage(
                            content.CallId ?? $"call_{Guid.NewGuid():N}",
                            content.Result ?? ""
                        ));
                    }
                    break;
            }
        }

        return openAIMessages;
    }

    private List<ChatTool>? ConvertToOpenAITools(ToolDefinition[]? protocolTools)
    {
        if (protocolTools == null || protocolTools.Length == 0)
        {
            return null;
        }

        var tools = new List<ChatTool>();

        foreach (var tool in protocolTools)
        {
            if (tool.Function == null) continue;

            var parametersJson = tool.Function.Parameters.HasValue
                ? BinaryData.FromString(tool.Function.Parameters.Value.GetRawText())
                : BinaryData.FromString("{}");

            tools.Add(ChatTool.CreateFunctionTool(
                tool.Function.Name,
                tool.Function.Description,
                parametersJson
            ));
        }

        return tools;
    }

    private class ToolCallBuilder
    {
        public string CallId { get; set; } = "";
        public string FunctionName { get; set; } = "";
        public string ArgumentsBuffer { get; set; } = "";
    }
}
